gapidapk/deviceinfo.go

```diff
@@ -156,6 +161,14 @@ func fetchDeviceInfo(ctx context.Context, d adb.Device) error {
 }
 
 func preparePerfettoProducerLauncherFromApk(ctx context.Context, d adb.Device) (string, app.Cleanup, error) {
+	if res, err := d.Shell("perfetto", "--query").Call(ctx); err == nil && (strings.Contains(res, "gpu.counters") || strings.Contains(res, "gpu.renderstages")) {
+		if pid, err := d.Shell("cat", perfettoProducerLauncherPidFile).Call(ctx); err == nil {
+			if _, err := strconv.Atoi(pid); err == nil {
+				d.Shell("kill -2", pid).Call(ctx)
+			}
+		}
+		return "", nil, log.Errf(ctx, err, "A producer is already running.")
```

Comment:
Here, a producer was running, but we just killed it. Why is it not OK to continue with the rest of the function?

Reply:
The previous approach causes an issue that if latter gapid instance is closed, then existing gapid instances will lose the producer because we always kill the previous existing producer.

So instead, I changed it to:

1. Check whether there are producers running;
2. If yes, abort existing producers and return, meaning not starting new one;
3. The previous gapid instance will get a nil return in this case, so have a for loop to restart producers.

Reply:
Does it mean that the `for true` loop is expected to loop until it actually hits an `err != nil`, at which point we are sure there is a producer running?